File: rxbinding-design/src/main/java/com/jakewharton/rxbinding2/support/design/widget/RxSnackbar.java
package com.jakewharton.rxbinding2.support.design.widget;

import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;
import android.support.design.widget.Snackbar;
import android.view.View;
import io.reactivex.Observable;

import static com.jakewharton.rxbinding2.internal.Preconditions.checkNotNull;

/**
 * Static factory methods for creating {@linkplain Observable observables} for {@link Snackbar}.
 */
public final class RxSnackbar {
  /**
   * Create an observable which emits the dismiss events from {@code view}.
   * <p>
   * <em>Warning:</em> The created observable keeps a strong reference to {@code view}. Unsubscribe
   * to free this reference.
   */
  @CheckResult @NonNull
  public static Observable<Integer> dismisses(@NonNull Snackbar view) {
    checkNotNull(view, "view == null");
    return new SnackbarDismissesObservable(view);
  }

  /**
   * Create an observable which emits the action click events from {@code view}.
   * <p>
   * <em>Warning:</em> The created observable keeps a strong reference to {@code view}. Unsubscribe
   * to free this reference.
   */
  @CheckResult @NonNull
  public static Observable<View> actionClicks(@NonNull Snackbar view, int resId) {
    checkNotNull(view, "view == null");
    return new SnackbarActionObservable(view, resId);
  }

  /**
   * Create an observable which emits the action click events from {@code view}.
   * <p>
   * <em>Warning:</em> The created observable keeps a strong reference to {@code view}. Unsubscribe
   * to free this reference.
   */
  @CheckResult @NonNull
  public static Observable<View> actionClicks(@NonNull Snackbar view, @NonNull CharSequence text) {
    checkNotNull(view, "view == null");
    checkNotNull(text, "text == null");
    return new SnackbarActionObservable(view, text);
  }

  private RxSnackbar() {
    throw new AssertionError("No instances.");
  }
}

File: rxbinding-design-kotlin/src/main/kotlin/com/jakewharton/rxbinding2/support/design/widget/RxSnackbar.kt
@file:Suppress(
    names = "NOTHING_TO_INLINE"
)

package com.jakewharton.rxbinding2.support.design.widget

import android.support.design.widget.Snackbar
import android.view.View
import io.reactivex.Observable
import kotlin.Int
import kotlin.Suppress

/**
 * Create an observable which emits the dismiss events from `view`.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 */
inline fun Snackbar.dismisses(): Observable<Int> = RxSnackbar.dismisses(this)

/**
 * Create an observable which emits the action click events from `view`.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 */
inline fun Snackbar.actionClicks(resId: Int): Observable<View> = RxSnackbar.actionClicks(this, resId)

/**
 * Create an observable which emits the action click events from `view`.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 */
inline fun Snackbar.actionClicks(text: CharSequence): Observable<View> = RxSnackbar.actionClicks(this, text)
